docs/book/src/tasks/experimental-features/runtime-sdk/implement-topology-mutation-hook.md

```diff
@@ -35,11 +38,103 @@ External patches have the following advantages:
 * External patches can use external data (e.g. from cloud APIs) during patch generation.
 * External patches can be easily reused across ClusterClasses.
 
+## External variables
+The DiscoverVariables hook can be used to supply variable definitions for use in external patches. These variable definitions are added to
+any applicable ClusterClasses. Clusters using the ClusterClass can then set values for those variables.
+
+### External variable discovery in the ClusterClass
+External variable definitions are discovered by calling the DiscoverVariables runtime hook. This hook is called from the ClusterClass reconciler.
+Once discovered the variable definitions are validated and stored in ClusterClass status.
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterClass
+# metadata
+spec:
+    # Inline variable definitions
+    variables:
+    - name: http-proxy
+      schema:
+        openAPIV3Schema:
+          type: string
+          default: "proxy.example.com"
+          example: "proxy.example.com"
+          description: "proxy for http calls."
+    # External patch definitions.
+    patches:
+    - name: lbImageRepository
+      external:
+          generateExtension: generate-patches.k8s-upgrade-with-runtimesdk
+          validateExtension: validate-topology.k8s-upgrade-with-runtimesdk
+          ## Explicitly call variable discovery for this patch.
+          variableDiscovery: variable-discovery.k8s-upgrade-with-runtimesdk
+status:
+    # observedGeneration is used to check that the current version of the ClusterClass is the same as that when the Status was previously written.
+    # if metadata.generation isn't the same as observedGeneration Cluster using the ClusterClass should not reconcile.
+    observedGeneration: xx
+    # variables contains a list of all variable definitions, both inline and external, that belong to the ClusterClass.
+    variables: 
+        - name: http-proxy
+          schema:
+            openAPIV3Schema:
+             type: string
+             # default is different so this variable does not match the inline definition.
+             default: "proxy.example3.com"
+             example: "proxy.example.com"
+             description: "proxy for http calls."
+          # namespace is added as this variable's schema isn't equivalent to the inline definition.
+          namespace: lbImageRepository
```

Comment:
I think namespace would fit better. I assume whatever we choose here is the same that users would use when defining the variable value?

origin seems a bit strange there

Reply:
Regarding: 1 entry per variable definition.

I think I would go with that. Otherwise the question is when do we merge entries together. Our current definition of "equivalent" wouldn't work as this means that we would have to drop description and example here.